packages/@aws-cdk/aws-codepipeline/lib/stage.ts

```diff
@@ -29,25 +41,37 @@ export class Stage extends cdk.Construct {
      */
     constructor(parent: Pipeline, name: string) {
         super(parent, name);
+        this.name = name;
         this.pipeline = parent;
-        validation.validateName('Stage', name);
+        actions.validateName('Stage', name);
+    }
+
+    public _addAction(action: actions.Action): void {
+        // _addAction should be idempotent in case a customer ever calls it directly
+        if (!this._actions.includes(action)) {
+            this._actions.push(action);
+        }
     }
 
     /**
      * Get a duplicate of this stage's list of actions.
      */
-    public get actions(): Action[] {
+    public get actions(): actions.Action[] {
         return this._actions.slice();
     }
 
     public validate(): string[] {
         return this.validateHasActions();
     }
 
```

Comment:
This looks like what I proposed earlier instead of another interface. I would allude to the bucket in the name of the method (because stage.grantReadWrite) is weird

Reply:
Yes, if you prefer (I hate long names). Since this method is part of the pipeline's stage, I think "pipeline" is not really needed by the context, but that's fine :-)